Sources/Target/POSIX/ELFProcess.cpp

```diff
@@ -53,43 +53,56 @@ inline ErrorCode GetELFSharedLibraryInfoAddress(ELFProcess *process,
   PHDR ph;
   DYNSYM dynsym;
 
-  CHK(process->readMemory(process->loadBase(), &eh, sizeof(eh)));
-
-  if (eh.e_phnum == 0) {
+  // 1. Find the PHDR, number of entries, size of
+  //    each entry, and the loadBase of the binary.
+  //    Assumption: PHDR is attached to the ELF
+  //    header. This isn't enforced but it almost
+  //    always occurs in practice.
+  uint64_t phdrAddress = process->getAuxiliaryVectorValue(AT_PHDR);
+  size_t phdrEntrySize = process->getAuxiliaryVectorValue(AT_PHENT);
+  size_t phdrNumEntry = process->getAuxiliaryVectorValue(AT_PHNUM);
+  uint64_t loadBase = phdrAddress - sizeof(eh);
+  if (phdrAddress == 0 || phdrEntrySize == 0 || phdrNumEntry == 0) {
     return kErrorUnsupported;
   }
 
-  // 1. Find PT_DYNAMIC (.dynamic) program header, and the
-  //    expected load base of this ELF.
+  // 2. Find PT_DYNAMIC (.dynamic) program header
+  //    and what the ELF believes is the loadBase.
+  //    Find the size of the dynamic section as well.
   uint64_t elfLoadBase = 0;
   bool foundElfLoadBase = false;
-  uint64_t phdrAddress = process->loadBase() + eh.e_phoff;
-  for (size_t n = 0; n < eh.e_phnum; n++) {
+  uint64_t dynamicAddress = 0;
+  uint64_t dynamicMemsz = 0;
+  bool foundDynamicAddress = false;
+  for (size_t n = 0; n < phdrNumEntry; n++) {
     CHK(process->readMemory(phdrAddress, &ph, sizeof(ph)));
 
-    if (!foundElfLoadBase && ph.p_type == PT_LOAD) {
-      elfLoadBase = ph.p_paddr;
+    if ((!foundElfLoadBase || ph.p_vaddr < elfLoadBase) &&
+        ph.p_type == PT_LOAD) {
+      elfLoadBase = ph.p_vaddr;
       foundElfLoadBase = true;
     }
 
     if (ph.p_type == PT_DYNAMIC) {
-      break;
+      dynamicAddress = ph.p_vaddr;
+      dynamicMemsz = ph.p_memsz;
+      foundDynamicAddress = true;
     }
 
-    phdrAddress += eh.e_phentsize;
+    phdrAddress += phdrEntrySize;
```

Comment:
Nit: Not your code, but I think changing the value of `phdrAddress` makes the code slightly harder to follow. Doing something like

    CHK(process->readMemory(phdrAddress + n * sizeof(ph), &ph, sizeof(ph)));

seems easier to read for me.